File: java/src/main/java/org/training/speedrun/stage5/Stage5.java
package org.training.speedrun.stage5;

/**
 * Make this code clean.
 */
public class Stage5 {

    double compute(double[] c, boolean v, boolean b) {
        double t = 0;
        for (int i = 0; i < c.length; ++i) {
            t += c[i];
        }
        int d = compute(c.length, b, v);
        double d2 = d * t / 100;
        return t + d2;
    }

    int compute(int n, boolean b1, boolean b2) {
        int p = 0;
        if (n >= 5 && n < 10) {
            p += 4;
        } else if (n >= 10) {
            p += 10;
        }
        if (b1) {
            // black friday discount
            p += 20;
        }

        if (b2) {
            // vip client
            p += 5;
        }
        return p;
    }
}

File: kotlin/src/main/kotlin/org/training/speedrun/stage5/Stage5.kt
package org.training.speedrun.stage5

import java.util.*

/**
 * Refactor this code until all methods body is <= 5 lines, and you find the code clean.
 */
class Stage5 {

    fun isEligible(word: String): Boolean {
        if (word.length >= 10) {
            val vowels = listOf('a', 'e', 'i', 'o', 'u', 'y')
            val nbVowels = word.filter { vowels.contains(it) }.count()
            if (nbVowels >= 4) {
                return true
            } else {
                val specialLetters = listOf('p', 'm')
                val nbSpecialLetters = word.filter { specialLetters.contains(it) }.count()
                if (nbSpecialLetters >= 3) {
                    return true
                }
            }
        } else {
            val nbLettersInUpperCase = word.chars().filter { c -> c >= 'A'.toInt() && c <= 'Z'.toInt() }.count()
            if (nbLettersInUpperCase >= 5) {
                return true
            } else {
                val vowels = listOf('a', 'e', 'i', 'o', 'u', 'y')
                val nbVowels = word.filter { vowels.contains(it) }.count()
                if (nbVowels >= 3) {
                    return true
                }
            }
        }
        return false
    }
}
